radicale/storage.py

```diff
@@ -474,3 +517,61 @@ def serialize(self):
         elif self.get_meta("tag") == "VADDRESSBOOK":
             return "".join([item.serialize() for item in items])
         return ""
+
+    _lock = threading.Lock()
+
+    @classmethod
+    def acquire_lock(cls, mode):
+        class Lock:
+            def __init__(self, release_method):
+                self._release_method = release_method
+
+            def release(self):
+                self._release_method()
+
+        if mode not in ("r", "w"):
+            raise ValueError("Invalid lock mode: %s" % mode)
+        folder = os.path.expanduser(
+            cls.configuration.get("storage", "filesystem_folder"))
+        if not os.path.exists(folder):
+            os.makedirs(folder, exist_ok=True)
+        lock_path = os.path.join(folder, "Radicale.lock")
+        lock_file = open(lock_path, "w+")
+        # set access rights to a necessary minimum to prevent locking by
+        # arbitrary users
+        try:
+            os.chmod(lock_path, stat.S_IWUSR | stat.S_IRUSR)
+        except OSError:
+            cls.logger.debug("Failed to set permissions on lock file")
+        locked = False
+        if os.name == "nt":
+            handle = msvcrt.get_osfhandle(lock_file.fileno())
+            flags = LOCKFILE_EXCLUSIVE_LOCK if mode == "w" else 0
+            overlapped = Overlapped()
+            if lock_file_ex(handle, flags, 0, 1, 0, overlapped):
+                locked = True
+        elif os.name == "posix":
+            operation = fcntl.LOCK_EX if mode == "w" else fcntl.LOCK_SH
+            # According to documentation flock() is emulated with fcntl() on
+            # some platforms. fcntl() locks are not associated with an open
+            # file descriptor. The same file can be locked multiple times
+            # within the same process and if any fd of the file is closed,
+            # all locks are released.
```

Comment:
Also the Linux manpage of fcntl doesn't indicate any different behavior: http://linux.die.net/man/2/fcntl

On 23 May 2016 09:36:35 CEST, Unrud <notifications@github.com> wrote:
>> +        except OSError:
>> +            cls.logger.debug("Failed to set permissions on lock
>file")
>> +        locked = False
>> +        if os.name == "nt":
>> +            handle = msvcrt.get_osfhandle(lock_file.fileno())
>> +            flags = LOCKFILE_EXCLUSIVE_LOCK if mode == "w" else 0
>> +            overlapped = Overlapped()
>> +            if lock_file_ex(handle, flags, 0, 1, 0, overlapped):
>> +                locked = True
>> +        elif os.name == "posix":
>> +            operation = fcntl.LOCK_EX if mode == "w" else
>fcntl.LOCK_SH
>> +            # According to documentation flock() is emulated with
>fcntl() on
>> +            # some platforms. fcntl() locks are not associated with
>an open
>> +            # file descriptor. The same file can be locked multiple
>times
>> +            # within the same process and if any fd of the file is
>closed,
>> +            # all locks are released.
>
>I think you misunderstood me. flock() and lockf() locks behave very
>different. Everything you said only applies to lockf() locks. flock()
>locks, which are used here, are not ineffective.
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/Kozea/Radicale/pull/402/files/ff3b31fed6d1c090e4957a4c3d9432a375f2d216#r64179106

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.